File: Java/Hard/30. Substring with Concatenation of All Words/30_Substring_with_Concatenation_of_All_Words.java
class Solution {
 public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || s.length() == 0 || words == null || words.length == 0) return result;

        int wordLength = words[0].length();
        int wordsCount = words.length;
        int substringLength = wordLength * wordsCount;

        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words) wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);

        for (int i = 0; i < wordLength; i++) {
            int left = i;
            Map<String, Integer> seen = new HashMap<>();
            int count = 0;

            for (int j = i; j <= s.length() - wordLength; j += wordLength) {
                String word = s.substring(j, j + wordLength);

                if (wordMap.containsKey(word)) {
                    seen.put(word, seen.getOrDefault(word, 0) + 1);
                    count++;

                    while (seen.get(word) > wordMap.get(word)) {
                        String leftWord = s.substring(left, left + wordLength);
                        seen.put(leftWord, seen.get(leftWord) - 1);
                        count--;
                        left += wordLength;
                    }

                    if (count == wordsCount) result.add(left);
                } else {
                    seen.clear();
                    count = 0;
                    left = j + wordLength;
                }
            }
        }
        return result;
    }
}

File: Golang/30. Substring with Concatenation of All Words/30_Substring_with_Concatenation_of_All_Words.go
func findSubstring(s string, words []string) []int {
    result := []int{}
    if len(s) == 0 || len(words) == 0 {
        return result
    }

    wordLength := len(words[0])
    wordsCount := len(words)


    wordMap := map[string]int{}
    for _, word := range words {
        wordMap[word]++
    }

    for i := 0; i < wordLength; i++ {
        left := i
        seen := map[string]int{}
        count := 0

        for j := i; j <= len(s)-wordLength; j += wordLength {
            word := s[j : j+wordLength]

            if _, ok := wordMap[word]; ok {
                seen[word]++
                count++

                for seen[word] > wordMap[word] {
                    leftWord := s[left : left+wordLength]
                    seen[leftWord]--
                    count--
                    left += wordLength
                }

                if count == wordsCount {
                    result = append(result, left)
                }
            } else {
                seen = map[string]int{}
                count = 0
                left = j + wordLength
            }
        }
    }
    return result
}
